syntax = "proto3";

import "google/protobuf/timestamp.proto";
import "google/protobuf/struct.proto";
import "user/user.proto";

service NotificationService {
  rpc get_notification_by_notification_id(NotificationRequestByNotificationId) returns (GrpcNotification){}
  rpc get_notifications_by_notifications_ids(NotificationsRequestByNotificationsIds) returns (ListOfNotifications){}
  rpc get_notifications_by_author_id(NotificationsRequestByAuthorId) returns (ListOfNotifications){}
  rpc get_notification_by_event_and_author_ids(NotificationRequestByEventAndAuthorIds) returns (GrpcNotification){}
  rpc get_all_notifications(GetAllNotificationsRequest) returns (ListOfNotifications){}
  rpc create_notification(NotificationRequest) returns (GrpcNotification){}
  rpc update_notification(NotificationRequest) returns (GrpcNotification){}
  rpc delete_notification_by_id(DeleteNotificationByIdRequest) returns (GrpcNotification){}
  rpc delete_notifications_by_events_and_author_ids(DeleteNotificationsByEventsAndAuthorIdsRequest) returns (ListOfNotifications){}
  rpc delete_notifications_by_event_id(DeleteNotificationsByEventIdRequest) returns (ListOfNotifications){}
  rpc delete_notifications_by_author_id(DeleteNotificationsByAuthorIdRequest) returns (ListOfNotifications){}
}

message NotificationsRequestByAuthorId {
  string author_id = 1;
  GrpcUser requesting_user = 2;
  uint64 page_number = 3;
  int64 items_per_page = 4;
}

message NotificationRequestByEventAndAuthorIds {
  string event_id = 1;
  string author_id = 2;
  GrpcUser requesting_user = 3;
}

message NotificationRequestByNotificationId {
  string notification_id = 1;
  GrpcUser requesting_user = 2;
}

message NotificationsRequestByNotificationsIds {
  ListOfIds notifications_ids = 1;
  GrpcUser requesting_user = 2;
  uint64 page_number = 3;
  int64 items_per_page = 4;
}

message GrpcNotification {
  string id = 1;
  string event_id = 2;
  string author_id = 3;
  bool enabled = 4;
  google.protobuf.Timestamp created_at = 5;
  oneof optional_deleted_at {
    google.protobuf.NullValue deleted_at_null = 6;
    google.protobuf.Timestamp deleted_at = 7;
  }
}

message GetAllNotificationsRequest {
  GrpcUser requesting_user = 1;
  uint64 page_number = 2;
  int64 items_per_page = 3;
}

message NotificationRequest {
  GrpcNotification notification = 1;
  GrpcUser requesting_user = 2;
}

message ListOfNotifications {
  repeated GrpcNotification notifications = 1;
}

message ListOfIds {
  repeated string ids = 1;
}

message DeleteNotificationByIdRequest {
  string notification_id = 1;
  GrpcUser requesting_user = 2;
}

message DeleteNotificationsByEventsAndAuthorIdsRequest {
  ListOfIds event_ids = 1;
  string author_id = 2;
  GrpcUser requesting_user = 3;
}

message DeleteNotificationsByEventIdRequest {
  string event_id = 1;
  GrpcUser requesting_user = 2;
}

message DeleteNotificationsByAuthorIdRequest {
  string author_id = 1;
  GrpcUser requesting_user = 2;
}
